### ISSUE TITLE
Confusing suggestion to use `cargo:rustc-link-lib` in a cargo project with a lib and bins

### ISSUE THREAD

**Issue by kusnezoff-alexander** [rust-lang/rust#129904]
### Problem

-- I created this repo to showcase/reproduce the error: https://github.com/kusnezoff-alexander/cargo-build-linking-bug --

The success of linking C-libraries during `cargo build` seems to depend on the existence of `lib.rs` even if only the binary is built.
Building packages which offer both a library and binary could suffer from this potential bug. This bug has been only been reproduced so far for linking against a library via
```rust
println!("cargo:rustc-link-search=../libs");
println!("cargo:rustc-link-lib=static=mylib");
```
inside `build.rs`.

**Exptected behavior**: The existence of a file called `lib.rs` should have no effect on the linking during `cargo build`:

**Observed behavior**:
1. Without the existence of `lib.rs`: Linking to custom C-library does work (see `./linking-without-lib-works` for working example)
2. Withthe existence of `lib.rs`: Linking to custom C-library does not work (see `./linking-with-lib-doesnt` for working example)

### Steps

1. `git clone https://github.com/kusnezoff-alexander/cargo-build-linking-bug`
2. `cd libs && make all`
3. `cd linking-does-work && cargo build` - should exit successfully
4. `cd linking-doesnt-work && cargo build` - should throw an error that extern function-symbol isn't defined (although the only difference to the previous case is the existence of lib.rs)

### Notes

- interestlingly though, it does work if `export RUSTFLAGS="-L <absolute-path-to-this-dir>/libs/ -l mylib"` is set

### Version

```text
cargo 1.80.1 (376290515 2024-07-16)
release: 1.80.1
commit-hash: 37629051518c3df9ac2c1744589362a02ecafa99
commit-date: 2024-07-16
host: x86_64-unknown-linux-gnu
libgit2: 1.7.2 (sys:0.18.3 vendored)
libcurl: 8.6.0-DEV (sys:0.4.72+curl-8.6.0 vendored ssl:OpenSSL/1.1.1w)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Fedora 40.0.0 [64-bit]
```

**Comment by ehuss** [comment]
Unfortunately I don't know of a feasible way to do that right now. rustc doesn't know anything about cargo, and it is the one generating the message. Cargo doesn't know anything about linking, and wouldn't be able to know what is happening.

Transferred to rust-lang/rust since this message is generated by rustc. I don't know if the wording could possibly be tweaked to avoid this confusion.
